### DISCUSSION TITLE
Provider: utm qemu libvirt

### DISCUSSION THREAD

**Discussion by georgestephanis** [Varying-Vagrant-Vagrants/VVV#2783] [General]
Was chatting with @evertiro in slack about this, logging some potentially useful links -- to the best of my knowledge, **all linked projects here are free and open source**.

UTM can run virtual machines beautifully on the M1 Macs using hypervisor -- https://mac.getutm.app/ -- it also has a version that runs on iPhones and iPads, but that's less salient to our purposes. https://github.com/utmapp/UTM/

There's been some questions on their repository about adding support for Vagrant:

https://github.com/utmapp/UTM/issues/2496

with some planned changes that may make it more doable:

https://github.com/utmapp/UTM/issues/2280

but it is noted there that UTM is largely a wrapper about QEMU.

UTM recently added some functionality to allow exporting the QEMU arguments out of UTM, if it simplifies anyone's workflow when doing initial explorations: https://github.com/utmapp/UTM/commit/48a8fc74fff07e12199670da2107fd4269a6d70f

QEMU -- https://www.qemu.org/ -- https://gitlab.com/qemu-project/qemu/ -- is a generic and open source machine emulator and virtualizer.

libvirt -- https://libvirt.org/ -- also works as a wrapper around qemu, and reportedly runs just fine on apple silicon! https://doesitarm.com/formula/libvirt/

It also has a vagrant provider: https://github.com/vagrant-libvirt/vagrant-libvirt

Anyway, just wanted to dump some starting points and research here for some assorted ways we could get VVV running smoothly on Apple Silicon with open source tools.

**Comment by welcome** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template and include your OS/Vagrant/VVV versions! Don't forget you can get support in the VVV slack at https://varyingvagrantvagrants.org/docs/en-US/slack/


<blockquote><img src="https://varyingvagrantvagrants.org/images/vvv.png" width="48" align="right"><div><img src="/images/vvv.png" height="14"> VVV</div><div><strong><a href="https://varyingvagrantvagrants.org/docs/en-US/slack/">VVV Slack Workspace</a></strong></div><div>Join the VVV Slack Workspace</div></blockquote>

**Comment by georgestephanis** [comment]
Possibly salient notes from someone else:

https://gist.github.com/aserhat/91c1d5633d395d45dc8e5ab12c6b4767

<blockquote><img src="https://github.githubassets.com/images/modules/gists/gist-og-image.png" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> Gist</div><div><strong><a href="https://gist.github.com/aserhat/91c1d5633d395d45dc8e5ab12c6b4767">QEMU and HVF</a></strong></div><div>QEMU and HVF. GitHub Gist: instantly share code, notes, and snippets.</div></blockquote>

**Comment by evertiro** [comment]
I don't think this is necessarily an M1 or even Mac-specific thing. QEMU/libvirt are available for all three supported operating systems, so if we can make it work this could be our first completely free cross-platform provider. I say it's worth a shot!

**Comment by evertiro** [comment]
I'd like to point out that libvirt already has a stable vagrant provider. I'd say that's probably the easiest place to start.

**Comment by tomjn** [comment]
For MacOS we ideally want something using Apples hypervisor, which is what VMWare Docker and Parallels use

**Comment by georgestephanis** [comment]
Based on what research I'd done, it seems qemu does also use hypervisor.

UTM explicitly says it does as well.

> UTM employs Apple's Hypervisor virtualization framework to run ARM64 operating systems on Apple Silicon at near native speeds.

and

> Under the hood of UTM is QEMU, a decades old, free and open source emulation software that is widely used and actively maintained.

According to https://www.naut.ca/blog/2020/08/26/ubuntu-vm-on-macos-with-libvirt-qemu/:

> `QEMU` on the other hand has had support for Hypervisor.Framework since 2018.

**Comment by tomjn** [comment]
then we need the following:

 - an arm64 Ubuntu 20 box
 - a provider plugin

This also assumes no efforts towards containerised VVV or native VVV

**Comment by tomjn** [comment]
This libvirt box looks the most promising at the moment on vagrant clouds box catalog:

https://app.vagrantup.com/travismccollum/boxes/u20_arm64

There's also xhyve, which has been around a while and uses a port of bhyve to interface with Apples hypervisor framework

https://github.com/machyve/xhyve

We would need to build our own box though, and I couldn't find any information in a brief look for Apple Silicon compatibility

<blockquote><div><img src="/favicon-32x32.png" height="14"> Vagrant Cloud by HashiCorp</div><div><strong><a href="https://app.vagrantup.com">Vagrant Cloud by HashiCorp</a></strong></div><div>Vagrant Cloud by HashiCorp</div></blockquote>
<blockquote><img src="https://opengraph.githubassets.com/815221dca24506d2aaa3bf6ca4f21daa24a2f0e1a6cabc38d353ec7ae8c30ab1/machyve/xhyve" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/machyve/xhyve">machyve/xhyve</a></strong></div><div>xhyve, a lightweight OS X virtualization solution. Contribute to machyve/xhyve development by creating an account on GitHub.</div></blockquote>

**Comment by tomjn** [comment]
If someone was willing to roll up their sleeves with xcode and some ruby, they could bypass all of these and build something specific to VVV+MacOS

This issue has a lot of interesting things in it https://github.com/machyve/xhyve/issues/206

**Comment by evertiro** [comment]
I mean eventually. Though I'm not normally a fan of writing ruby.

**Comment by tomjn** [comment]
[Multipass](https://multipass.run/) by canonical may be a viable alternative, it uses Apples hypervisor behind the scenes on MacOS, kvm on linux, and hyper-v on Windows. There's a [Vagrant plugin](https://github.com/Fred78290/vagrant-multipass) but it hasn't been touched in 14 months.

We would need to figure out our own box system though, or if boxes even make sense for multipass, there's nothing in a search for multipass on vagrant cloud

**Comment by GregOriol** [comment]
I'm very interested in this: I'd like to run an ubuntu x86 via vagrant on an M1 mac. The guest system runs with UTM/qemu, what is missing is the interaction with vagrant. It would be very useful in that setup because some packages I use inside the vagrant box (which is similar to the production environment) don't have available binaries for ubuntu arm...

**Comment by tomjn** [comment]
It may be possible to build a primitive rigged vagrant plugin by building a VM in UTM that can be downloaded that's preconfigured then using `utm://` links but it'll be hacky and probably have a GUI window:

https://github.com/utmapp/UTM/wiki/URL-Scheme-%E2%80%93-Automation-for-UTM-app-&-VMs

Also relevant: https://github.com/utmapp/UTM/discussions/3618

Possibly require us to mess with their plist format, it would require a bit of effort but it looks doable. The concern I have is shared/mounted folders may be a pain in the arse to figure out.

What I would like is someone who can do the following:

 - setup a UTM virtual machine on Apple Silicon running headless Ubuntu 20 with 2 or 3 shared folders
 - export that VM somewhere, including any plist files from UTM for inspection

This way we can try to reverse engineer some things and figure stuff out

cc @GregOriol @Mamaduka

**Comment by trinitronx** [comment]
> libvirt -- https://libvirt.org/ -- also works as a wrapper around qemu, and reportedly runs just fine on apple silicon! https://doesitarm.com/formula/libvirt/
> 
> It also has a vagrant provider: https://github.com/vagrant-libvirt/vagrant-libvirt

This is [quite possible to do right now with fully native `arm64` `libvirt` & `qemu` binaries + gem native `arm64` extensions][1].  The [main obstacle is getting `vagrant` itself to run as a gem][2] on the Apple Silicon M1, because [there is no official `arm64` release yet][3].

If using macOS VM guests ([yes, it's possible][5], but there's some tricks to it... e.g. they need [NVRAM][6], and some [magic incantations][7]. It also requires [a few more unreleased gem patches][8].)

[Status update on all the dependencies needed to get this working with the `libvirt` provider can be found here][9]

EDIT: The other nice feature with using `qemu` + `libvirt` is that [you can launch VMs on a remote host machine over the network with `qemu+ssh://` protocol][10].

[1]: https://github.com/vagrant-libvirt/vagrant-libvirt/issues/1205#issuecomment-1020261595
[2]: https://github.com/vagrant-libvirt/vagrant-libvirt/issues/1205#issuecomment-1018029758
[3]: https://github.com/hashicorp/vagrant/issues/12548
[4]: https://github.com/vagrant-libvirt/vagrant-libvirt/issues/1205#issuecomment-1083794064
[5]: https://github.com/kholia/OSX-KVM/
[6]: https://github.com/LyraPhase/vagrant-macos-monterey-12-1-base/blob/master/Vagrantfile#L114-L115
[7]: https://github.com/LyraPhase/vagrant-macos-monterey-12-1-base/blob/master/Vagrantfile#L58-L135
[8]: https://github.com/vagrant-libvirt/vagrant-libvirt/issues/1371#issuecomment-1120223051
[9]: https://github.com/vagrant-libvirt/vagrant-libvirt/issues/1205#issuecomment-1083794064
[10]: https://github.com/LyraPhase/sprout-wrap/blob/master/test/fixtures/Vagrantfile.libvirt-ssh#L14

**Comment by tomjn** [comment]
@trinitronx vagrant runs on Apple silicon, I and others have been using M1 type machines with Parallels for months.

> If using macOS VM guests ([yes, it's possible](https://github.com/kholia/OSX-KVM/), but there's some tricks to it... e.g. they need [NVRAM](https://github.com/LyraPhase/vagrant-macos-monterey-12-1-base/blob/master/Vagrantfile#L114-L115), and some [magic incantations](https://github.com/LyraPhase/vagrant-macos-monterey-12-1-base/blob/master/Vagrantfile#L58-L135). It also requires https://github.com/vagrant-libvirt/vagrant-libvirt/issues/1371#issuecomment-1120223051.)

VVV uses debian/Ubuntu type VMs, we aren't interested in MacOS guests.

> because https://github.com/hashicorp/vagrant/issues/12548.

I don't see much progress here, especially with Vagrant being ported to Golang. Arm64 builds on golang should be significantly easier to get going, and I expect any Arm support vagrant provides to happen there first.

It makes dealing with this that much more difficult as writing a UTM provider for example means writing it in ruby with all the deprecations coming in 2.3 and 2.4, or, writing it in golang and having to wait for 2.3 for limited documentation to be released.

---

Another primary problem is that if we send users to install VirtualBox they get an installer, run it, reboot, done. But if we send them to install QEMU it's much more difficult. That's why we hoped UTM would resolve the issue. A lot of VVV users are not technical, don't know how to program, and struggle with the most basic aspects of CLI usage.

E.g. if we send users to libvirt.org or just say "you need libvirt" then ¯\_(ツ)_/¯ we won't get any users, it's gibberish, years beyond the technical skill of 99% of the users who would need such a provider, making the whole endeavour pointless.

**Comment by georgestephanis** [comment]
Just noting that I'd spotted this today as a possible way of letting vagrant use qemu directly as a vagrant plugin

https://github.com/ppggff/vagrant-qemu

<blockquote><img src="https://opengraph.githubassets.com/5de3a865a91dccf054d14cb12adae76c80759cd04cde44704647f08d7dd9bea0/ppggff/vagrant-qemu" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/ppggff/vagrant-qemu">GitHub - ppggff/vagrant-qemu: Use Vagrant to manage machines using QEMU. Test with Apple Silicon / M1 and CentOS aarch64 image</a></strong></div><div>Use Vagrant to manage machines using QEMU. Test with Apple Silicon / M1 and CentOS aarch64 image - GitHub - ppggff/vagrant-qemu: Use Vagrant to manage machines using QEMU. Test with Apple Silicon /...</div></blockquote>

**Comment by jasalt** [comment]
Looking to get rid of VirtualBox on Debian 13 (amd64) as the version in distro package repository never works and gave libvirt/QEMU/KVM a go provisioning VVV. 

Here's `Vagrantfile` changes with minimally working setup https://github.com/jasalt/VVV/commit/dbad3795323a2802118967faa5749c498bacd46c.

1) Installed packages (Debian 13):
```
sudo apt install vagrant vagrant-libvirt libvirt-daemon-system virtiofsd qemu-utils virt-manager
```

2) Created a pool named default (fails if doesn't exist):
```
sudo virsh -c qemu:///system pool-define-as default dir - - - - /var/lib/libvirt/images
sudo virsh -c qemu:///system pool-build default
sudo virsh -c qemu:///system pool-start default
sudo virsh -c qemu:///system pool-autostart default
```

3) Startup:
```
vagrant up --provider=libvirt
```

<img width="1332" height="1031" alt="Image" src="https://github.com/user-attachments/assets/d5cf92e1-4992-405d-a9dd-1b4e0158022b" />

Networking and folder sharing seems to work and continuing testing it out. Also looking to automate the step 2) if the setup seems useful otherwise.

No idea how this works or makes sense on other platforms (e.g. MacOS) but maybe it can be useful to Linux users.

**Comment by tomjn** [comment]
@jasalt I've created a PR for this at https://github.com/Varying-Vagrant-Vagrants/VVV/pull/2780, but there's a few things needed for it:

 - changelog update
 - Ideally it would print some sort of warning if libvirt is used indicating it's not a mature/stable provider yet
 - If it's not possible to automate running those commands, could we detect if they need to be ran, or share them in the output?
   - How do other libvirt setups get around this? I'm not familiar with QEMU/libvirt so I'd be unsure how to get this installed and what those commands/pools do

**Comment by jasalt** [comment]
@tomjn Nice. These additions would make sense. 

Commenting on the PR https://github.com/Varying-Vagrant-Vagrants/VVV/pull/2780 on how it goes.
